Sources/Drill/DiagnosticVerifier.swift

```diff
@@ -0,0 +1,188 @@
+  public func verify() {
+    // Keep a list of expectations we haven't matched yet.
+    var unmatchedExpectations = expectedDiagnostics
+
+    // Maintain a list of unexpected diagnostics and the line they
+    // occurred on.
+    var unexpectedDiagnostics = [Int: Diagnostic]()
+
+    // Go through each diagnostic we've produced.
+    for diagnostic in producedDiagnostics {
+
+      // Expectations require a line.
+      guard let node = diagnostic.node, let loc = node.startLoc else {
+        engine.diagnose(.diagnosticWithNoNode(diagnostic.message))
+        continue
+      }
+
+      let expectation = Expectation(message: diagnostic.message,
+                                    tokenContainingComment: node)
+
+      // Make sure we're expecting this diagnostic.
+      guard expectedDiagnostics.contains(expectation) else {
+        unexpectedDiagnostics[loc.line] = diagnostic
+        continue
+      }
+
+      // Remove it from the unmatched one, if we've seen it.
+      unmatchedExpectations.remove(expectation)
+    }
+
+    // Diagnostics we never saw produced are errors -- make our own error
+    // stating that.
+    let expectations = unmatchedExpectations.sorted { a, b in
+      guard let aLine = a.tokenContainingComment.startLoc?.line,
+        let bLine = b.tokenContainingComment.startLoc?.line else {
+          return false
+      }
+      return aLine < bLine
+    }
+    for expectation in expectations {
+      if
+        let line = expectation.tokenContainingComment.startLoc?.line,
+        let unexpected = unexpectedDiagnostics[line] {
+        engine.diagnose(.incorrectDiagnostic(got: unexpected.message),
+                        node: unexpected.node) {
+          $0.note(.expected(expectation.message), node: unexpected.node)
+        }
+        unexpectedDiagnostics.removeValue(forKey: line)
+      } else {
+        engine.diagnose(.diagnosticNotRaised(expectation.message),
+                        node: expectation.tokenContainingComment)
+      }
+    }
+    for diagnostic in unexpectedDiagnostics.values {
+      engine.diagnose(.unexpectedDiagnostic(diagnostic.message),
+                      node: diagnostic.node)
+    }
+  }
+
+  /// Extracts a list of expectations from the comment trivia inside the
+  /// provided token stream. This provides the set of expectations that the
+  /// real diagnostics will be verified against.
+  private static func parseExpectations(
+    _ tokens: [TokenSyntax], engine: DiagnosticEngine) -> Set<Expectation> {
+    var expectations = Set<Expectation>()
+    for token in tokens {
+      for trivia in token.leadingTrivia + token.trailingTrivia {
+        guard case .comment(let text) = trivia else { continue }
+
+        let nsString = NSString(string: text)
+        let range = NSRange(location: 0, length: nsString.length)
+
+        diagRegex.enumerateMatches(in: text,
+                                   range: range) { result, _, _ in
+          guard let result = result else { return }
```

Comment:
Nit: Guard on the number of ranges being valid too.

Reply:
👍